data-analytics/iot-nirvana/app-engine/src/main/java/com/google/cloud/demo/iot/nirvana/shared/Utils.java

```diff
@@ -0,0 +1,373 @@
+ * the License.
+ */
+
+package com.google.cloud.demo.iot.nirvana.shared;
+
+import com.github.rholder.retry.RetryException;
+import com.github.rholder.retry.Retryer;
+import com.github.rholder.retry.RetryerBuilder;
+import com.github.rholder.retry.StopStrategies;
+import com.github.rholder.retry.WaitStrategies;
+import com.google.appengine.api.taskqueue.Queue;
+import com.google.appengine.api.taskqueue.QueueFactory;
+import com.google.appengine.api.taskqueue.TaskOptions;
+import com.google.appengine.api.taskqueue.TaskOptions.Method;
+import com.google.appengine.api.utils.SystemProperty;
+import com.google.common.base.Throwables;
+import com.google.gson.Gson;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+import com.google.gson.reflect.TypeToken;
+import java.io.BufferedReader;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.lang.reflect.Type;
+import java.security.MessageDigest;
+import java.util.Enumeration;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Properties;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Logger;
+import javax.servlet.ServletContext;
+import javax.servlet.http.Cookie;
+import javax.servlet.http.HttpServletRequest;
+
+/** Class that describes functions used throughout the application */
+public class Utils {
+
+  private static final Logger LOG = Logger.getLogger(Utils.class.getName());
+
+  /**
+   * Convert an object to JSON
+   *
+   * @param obj the object to be encoded
+   * @return encoded object
+   */
+  public static String toJSON(Object obj) {
+    Gson gson = new Gson();
+    return gson.toJson(obj);
+  }
+
+  /**
+   * Convert a JSON string to an Object
+   *
+   * @param json the json to be converted
+   * @param destinationClass the object to be generated
+   * @return encoded object
+   */
+  public static <T> T fromJSON(String json, Class<T> destinationClass) {
+    Gson gson = new Gson();
+    return gson.fromJson(json, destinationClass);
+  }
+
+  /**
+   * Check if the app is running in production or in local dev
+   *
+   * @return true if it is in production, false otherwise
+   */
+  public static boolean isGaeProduction() {
+    return SystemProperty.environment.value() == SystemProperty.Environment.Value.Production;
+  }
+
+  /**
+   * Check if the running environment is the dev
+   *
+   * @return true if it is the dev, false otherwise
+   */
+  public static boolean isGaeDevEnvironment() {
+    return !isGaeProduction();
+  }
+
+  /**
+   * Get input parameters if they are stored in JSON format
+   *
+   * @param req
+   * @return
+   * @throws IOException
+   */
+  public static Map<String, String> getRequestParameters(HttpServletRequest req)
+      throws IOException {
+    // read input data
+    InputStream is = req.getInputStream();
```

Comment:
String inputParams = new Scanner( is ).useDelimiter("\\A").next();

This would be sufficient to read the input request as string. Even better, IOUtils can be used to read all the stream as String.

Reply:
Changed to using IOUtils in commit https://github.com/GoogleCloudPlatform/professional-services/pull/61/commits/bdfede8409eb6660d48b75bcdb1897bad3757c02.